### PR TITLE
[AIRFLOW-6135] Extract DAG processing from SchedulerJob into separate class

### PR DESCRIPTION
**Pull request by mik-laj**
Hello,

Currently, ScheduleJob class has two uses cases:

as a scheduler loop - started as the main process
used for DAG processing - started as many separate processes controlled by 
The logic of this code is completely independent. For this reason, it is worth separating this class to make it easier to understand the code processing flow. 

Now this code is very misleading, because we can get the impression that the main process is responsible for more work than it actually does.

Now I will work on updating the tests, but the code works fine.

Table with the way of class division

| Name of method                              | Target class     |
|---------------------------------------------|------------------|
| _exit_gracefully                            | SchedulerJob     |
| is_alive                                    | SchedulerJob     |
| manage_slas                                 | DagFileProcessor |
| update_import_errors                        | DagFileProcessor |
| create_dag_run                              | DagFileProcessor |
| _process_task_instances                     | DagFileProcessor |
| _change_state_for_tis_without_dagrun        | SchedulerJob     |
| __get_concurrency_maps                      | SchedulerJob     |
| _find_executable_task_instances             | SchedulerJob     |
| _change_state_for_executable_task_instances | SchedulerJob     |
| _enqueue_task_instances_with_queued_state   | SchedulerJob     |
| _execute_task_instances                     | SchedulerJob     |
| _change_state_for_tasks_failed_to_execute   | SchedulerJob     |
| _process_dags                               | DagFileProcessor |
| _process_executor_events                    | SchedulerJob     |
| _execute                                    | SchedulerJob     |
| _execute_helper                             | SchedulerJob     |
| _find_dags_to_process                       | DagFileProcessor |
| process_file                                | DagFileProcessor |
| heartbeat_callback                          | SchedulerJob     |

The method code has not changed.

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6135
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by mik-laj** [comment]
@ashb @potiuk @kaxil @nuclearpinguin It works for me. Can you look at it?

**Comment by potiuk** [comment]
Yeah. Seems we have a lot of big changes to the code recently. I will take a look shortly.

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -54,8 +54,9 @@
 from airflow.utils.state import State
 
 
-class DagFileProcessor(AbstractDagFileProcessor, LoggingMixin):
-    """Helps call SchedulerJob.process_file() in a separate process.
+class DagFileProcessorRunner(AbstractDagFileProcessorRunner, LoggingMixin):
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: Do we really need a separate class for DagFileProcessorRunner and DagFileProcessor -- the "Runner" version doesn't seem like it adds much (they are always created together in the same process) so I think I'd prefer this as just a single class.

File: airflow/jobs/__init__.py [hunk 2]
```diff
@@ -23,4 +23,4 @@
 from airflow.jobs.backfill_job import BackfillJob  # noqa: F401
 from airflow.jobs.base_job import BaseJob  # noqa: F401
 from airflow.jobs.local_task_job import LocalTaskJob  # noqa: F401
-from airflow.jobs.scheduler_job import DagFileProcessor, SchedulerJob  # noqa: F401
+from airflow.jobs.scheduler_job import DagFileProcessor, DagFileProcessorRunner, SchedulerJob  # noqa: F401
```
Comment (ashb) [inline airflow/jobs/__init__.py]: I would say we shouldn't export this here -- it's not a job and not needed for any of the rest of the jobs.

**Review by ashb — COMMENTED** [review]
OMG yes, thank you. The back-and-forth/in-and-out of SchedulerJob and DagProcessorManager was confusing to follow.

I think that the `DagFileProcessor` etc classes (how ever we decide to split them) would be better placed in utils.dag_processing -- that way we could probably also then remove the Abstract* versions -- there is only a single implementation of them anyway.

**Comment by ashb** [comment]
(Will likely conflict with #6596)

**Comment by potiuk** [comment]
> (Will likely conflict with #6596)

Yep. That's why I think it's better to review this one after it is rebased by @mik-laj after my change gets merged.

File: airflow/jobs/scheduler_job.py (hunk 1 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: I would like to separate classes that are executed in different threads and processes. New process = New class. This will greatly facilitate the understanding of the code. If we combine it, there will be a lot of methods available, but some intended only for execution from the main process  and others from the child process.

File: airflow/jobs/__init__.py (hunk 2 above)
Comment (mik-laj) [reply airflow/jobs/__init__.py]: Ok. Agree. I will remove it.

**Comment by mik-laj** [comment]
I would love to wait for Jarek's changes. I am pleased with the warm words about this change.

File: airflow/jobs/scheduler_job.py (hunk 1 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: I think the name DagFileProcessorProcess will be better. This better describes its roles.

Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: I like the idea: _**New process = New class**_

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -290,105 +292,28 @@ def start_time(self):
...
-        """
-        if grace_multiplier is not None:
-            # Accept the same behaviour as superclass
-            return super().is_alive(grace_multiplier=grace_multiplier)
-        scheduler_health_check_threshold = conf.getint('scheduler', 'scheduler_health_check_threshold')
-        return (
-            self.state == State.RUNNING and
-            (timezone.utcnow() - self.latest_heartbeat).seconds < scheduler_health_check_threshold
-        )
+    :param dag_ids: If specified, only look at these DAG ID's
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py:314]:
```suggestion

    :param dag_ids: If specified, only look at these DAG ID's
```

File: airflow/jobs/scheduler_job.py [hunk 4]
```diff
@@ -290,105 +292,28 @@ def start_time(self):
...
+    :type dag_ids: list[unicode]
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py]:
```suggestion
    :type dag_ids: list[str]
```

now that we no longer support Py2

**Review by kaxil — APPROVED** [review]
Good Job @mik-laj . 

minor changes suggested. LGTM otherwise

**Comment by kaxil** [comment]
Just read Ash's comment and I think it might be good to remove Abstract* versions  !

>I think that the DagFileProcessor etc classes (how ever we decide to split them) would be better placed in utils.dag_processing -- that way we could probably also then remove the Abstract* versions -- there is only a single implementation of them anyway.

**Comment by mik-laj** [comment]
@Kaxil I agree, but for now I am trying to solve one problem. This change is already very big - +2,256 −2,169  In the near future I will probably make other changes in the schedule to improve it.  This is one of the first because customers complain about problems with it.

**Comment by kaxil** [comment]
> @kaxil I agree, but for now I am trying to solve one problem. This change is already very big - +2,256 −2,169 In the near future I will probably make other changes in the schedule to improve it. This is one of the first because customers complain about problems with it.

Sure, I am definitely fine with it :) 

Keep the PRs flowing, you have been doing a really good work

**Comment by mik-laj** [comment]
@ashb @kaxil  @potiuk  I recreated this PR because I had too many conflicts. Can i ask for merge this change?

File: airflow/jobs/scheduler_job.py [hunk 5]
```diff
@@ -293,105 +294,31 @@ def start_time(self):
...
-        if grace_multiplier is not None:
-            # Accept the same behaviour as superclass
-            return super().is_alive(grace_multiplier=grace_multiplier)
-        scheduler_health_check_threshold = conf.getint('scheduler', 'scheduler_health_check_threshold')
-        return (
-            self.state == State.RUNNING and
-            (timezone.utcnow() - self.latest_heartbeat).seconds < scheduler_health_check_threshold
-        )
+    :param dag_ids: If specified, only look at these DAG ID's
+    :type dag_ids: List[str]
```
Comment (potiuk) [inline airflow/jobs/scheduler_job.py:315]: What I think should happen here as well is to add type annotations and make those classes pylint compliant (and remove --coding--)? 

It should likely be done in a separate PR. I am working on such PR :)

**Review by potiuk — APPROVED** [review]

File: airflow/jobs/scheduler_job.py (hunk 5 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py:315]: I agree. We should harden the core and this is the first change that does it. Gradually, I will try to introduce further improvements that do this.
